### ISSUE TITLE
Is there anyway to keep a widget state even after a kernel restart?

### ISSUE THREAD

**Issue by andyjmc** [jupyter-widgets/ipywidgets#2369]
I have an issue. If I re-start my kernel then it will unlink my widgets (I get the little red broken chain symbol). Is there anyway to save the widget state in the notebook so even if I move that notebook to another kernel the widgets will remember their state? (saving the state json to a file maybe?)

**Comment by jasongrout** [comment]
You can save the widget state to the notebook (in classic notebook, see the Widget menu). However, it is basically just saving a snapshot. We can't reconstitute those widgets in a new kernel.

**Comment by kfk** [comment]
Not sure if this makes sense, but I am saving state in a json file. My business case is interactive sql query building so in that case once the user has selected schemas, tables and columns I save all in a json so that that sql can be rebuilt next time from the json. Also, I use the json to repopulate the widget values on each new run time.

**Comment by chaffra** [comment]
Can this be done from the widget object like `widget.to_json()` and then `widget.from_json()`?

**Comment by chaffra** [comment]
> Not sure if this makes sense, but I am saving state in a json file. My business case is interactive sql query building so in that case once the user has selected schemas, tables and columns I save all in a json so that that sql can be rebuilt next time from the json. Also, I use the json to repopulate the widget values on each new run time.

How do you save with widget state to a json file? I cannot find a function for that.

**Comment by hainm** [comment]
@chaffra  there is a notebook option for it: https://ipywidgets.readthedocs.io/en/latest/embedding.html

> The Download Widget State option triggers the downloading of a JSON file containing the serialized state of all the widget models currently in use,

**Comment by chaffra** [comment]
> @chaffra there is a notebook option for it: https://ipywidgets.readthedocs.io/en/latest/embedding.html
> 
> > The Download Widget State option triggers the downloading of a JSON file containing the serialized state of all the widget models currently in use,

@hainm , yes but I was looking more for the python function I could save to json and then reload the values into the widget from json.

**Comment by jasongrout** [comment]
Perhaps https://ipywidgets.readthedocs.io/en/latest/embedding.html#python-interface from the ipywidgets docs is helpful here?

**Comment by chaffra** [comment]
Thanks, `embed_data` can be used to save the state but how do I load the value back into the widget. I tried `widget.set_state(state)` unsuccessfully.

**Comment by jasongrout** [comment]
I'm not sure exactly what you are trying to do, but this may be trickier than it seems at first glance. The embed data here is the state of all widgets in the widget manager. Each widget has a unique id. Separately you have the code variables in your code, and we can't keep track of those - that's outside of the widget's knowledge.

Reconstituting object state probably involves matching those variable ids up to variables in the code. For example, suppose you have two sliders, with internal ids 'A123' and 'B456'. Your widget state here just says that slider A123 has these values, and slider B456 has these values. To reconstitute that into widgets, presumably you'll know you need two sliders, so you'll create two. But now you need to know which of these sliders you should reset with A123's values, and which slider you should reset with B456's values.

Another way to do this for a much simpler case is to just try to restore a single top-level widget. You ask for the state (recursively) of that one widget, and get back a json blob (I don't think there is a nice convenience function for this yet). Then there is a new function (that also doesn't exist yet), that can take that json blob, reconstitute the widgets, and hand you back the top-level widget, which you can then assign to a variable.

**Comment by jasongrout** [comment]
Perhaps it would be helpful if someone wrote down a very specific usecase, along with exactly the code they would like to work, with an explanation of what that code should do. Then if something like that is possible, we can help more clearly. If it is not possible currently, then that informs whoever wants to implement something like this.

**Comment by chaffra** [comment]
Thanks for the thourough answer. If one could choose the model_id for each
widget when it's created then the first solution might work. One would just
have to compare the key in the stored (json) widget state to the recreated
widget model_id. If the key matches the model_id the the stored value for
that key is used to update the widget.

If this sounds reasonable I could try implementing a recursive function
like that.

**Comment by pmavrodiev** [comment]
I mean what's the point in having the option to save the widget state when there's no way to load it back?

**Comment by sindhu51188** [comment]
Y

**Comment by jasongrout** [comment]
> I mean what's the point in having the option to save the widget state when there's no way to load it back?

The purpose of saving the widget state is to provide a representation of the widgets in a document not connected to original kernel (for example, if the notebook is exported to html like our documentation, or as a preview of outputs that will be created when cells are run).

**Comment by JonathanVHoecke** [comment]
This is discussion is really relevant and appears to be still open. Would there be any progress in the development of saving widgets' state and reloading these into the notebook?
Thank you for your comments

**Comment by jasongrout** [comment]
I don't know of any conversation beyond the conversation above about this topic.



> Would there be any progress in the development of saving widgets' state and reloading these into the notebook?

Note that widget state already can be saved in the notebook, and the corresponding widgets are reloaded onto the page when the notebook is opened. What is not done is reconstitute those saved widgets in the kernel as "live" widgets.

**Comment by JonathanVHoecke** [comment]
Dear Jason
Thank you for your reply, I am not sure if I understood what you meant with "note ,,, widgets are reloaded". Do you have a link to an example to what you mean with that? Thank you very much

**Comment by jasongrout** [comment]
> I am not sure if I understood what you meant with "note ,,, widgets are reloaded"

In JupyterLab, if you have the Setting > Save Widget State Automatically checked, then any live widgets in the document are saved in the document when it is saved. If you close the notebook and shut down the kernel, then reopen the document, those widgets will appear. However, they will not be "live" in that they are not hooked up to anything in the kernel, but they still will be displayed.

**Comment by stefaneidelloth** [comment]
Here is a use case:

I would like to allow my users to select a template for the notebook. Each template represents some colors, sizes, font type etc. 

a) If the user re-executes the full notebook, I want that the previously selected template is kept instead of falling back to the default template. I created a custom template widget to allow the user to easily select the template with a DropDown. 

![image](https://user-images.githubusercontent.com/12675339/154026915-7f4dba31-56c2-4d0b-9a84-5ca10fe23a59.png)

b) A workaround would be to use a code cell, and force the user to adapt the code to make the selection: 

![image](https://user-images.githubusercontent.com/12675339/154027270-190685af-f982-4e5c-809a-3321c546bec9.png)

When re-running that code cell, the selected/wanted template would still be valid. 

c) In a mixed approach I could pass a default value to the template control. However, that would force the user to change the code to get a permanent selection. 

d) Maybe it would be possible to store the widget state in terms of **tags of the cell containing the widget**. When the widget is created it checks if its parent cell already has some tags, determining the initial state of the widget. 

=> Does option d) make sense for you?

**Comment by jklap** [comment]
> In JupyterLab, if you have the Setting > Save Widget State Automatically checked, then any live widgets in the document are saved in the document when it is saved. If you close the notebook and shut down the kernel, then reopen the document, those widgets will appear. However, they will not be "live" in that they are not hooked up to anything in the kernel, but they still will be displayed.

This isn't what I'm seeing currently in 3.4.7. The state is saved in the notebook (verified that) but when I kill the kernel and re-open the notebook I'm getting "Error displaying widget: model not found" -- it seems as though display is querying the new kernel and not using any saved/embedded state?

**Comment by pbcornelius** [comment]
I can replicate this with a fresh conda Python 3.10 environment:

```
jupyterlab                3.5.3
ipywidgets                8.0.4
```

Automatically save widget state enabled.

Create new Jupyter notebook in Jupyter lab with two cells:
```
import ipywidgets as widgets
```

```
widgets.Label('Hello World!')
```

Close & open Notebook (new kernel): "Error displaying widget: model not found".

Open *same* notebook in Jupyter Notebooks: label is rendered.

**Comment by creative-resort** [comment]
This is an important user experience issue to me, too.

After instantiation, user input becomes available to the kernel, even 'continuous update' is possible.
Saved user input could also be conveyed the same way to the kernel upon object creation, so the widget's value is initialized with saved data.

An explicit initialization should be avoided by the user in that case, of course.
